frontend/components/common/InstantSearchProvider.tsx

```diff
@@ -96,20 +103,43 @@ export function InstantSearchProvider({
                }),
             }}
          >
+            <RefreshSearchResults apiKey={algoliaApiKey} />
             {children}
          </InstantSearch>
       </InstantSearchSSRProvider>
    );
-   if (routing) {
-      <InstantSearchSSRProvider {...serverState}>
-         {provider}
-      </InstantSearchSSRProvider>;
-   }
-   return provider;
 }
 
 function useCountRenders() {
    const countRef = React.useRef(0);
    countRef.current++;
    return countRef.current;
 }
+
+type RefreshSearchResultsProps = {
+   apiKey: ApiKey;
+};
+
+/**
+ * Returns a component that takes in an API key and refreshes search results
+ * if the API key changes (but only refreshes once for each api key change.
+ */
+function useSearchRefresher() {
+   // Create the ref in the parent component so it can live across re-renders
+   const previousApiKeyRef = React.useRef<ApiKey|null>(null);
+
+   return function RefreshSearchResults({
+      apiKey,
+   }: RefreshSearchResultsProps): null {
+      const { refresh } = useSearchCache();
+      useSafeLayoutEffect(() => {
+         const isFirstRender = !previousApiKeyRef.current;
+         const hasApiKeyChanged = previousApiKeyRef.current !== apiKey;
+         if (hasApiKeyChanged && !isFirstRender) {
+            refresh();
+         }
+         previousApiKeyRef.current = apiKey;
+      }, [apiKey, refresh]);
+      return null;
+   };
+}
```

Comment:
Returning a component declaration from a hook, makes the React reconciler sees it as a new component on every render (since the function reference changes). We can achieve the same result without the hook, by accepting the previous Algolia key as a prop:

```suggestion
type RefreshSearchResults = { apiKey: ApiKey; previousApiKey: ApiKey };

function RefreshSearchResults({ apiKey, previousApiKey }: RefreshCacheProps) {
   const { refresh } = useSearchCache();
   useSafeLayoutEffect(() => {
      const isFirstRender = previousApiKey == null;
      if (previousApiKey !== apiKey && !isFirstRender) {
         refresh();
      }
   }, [apiKey, previousApiKey, refresh]);
   return null;
}

```

Reply:
Unfortunately when `InstantSearch` unmounts, there's nothing you can do to prevent all of its children from being unmounted. These are what the React doc says:

<img width="964" alt="image" src="https://user-images.githubusercontent.com/4640135/176624950-f9596cef-ebd4-485b-805e-bdf81b9dab01.png">

In this particular scenario, returning a component from a hook doesn't change much, since `InstantSearch` unmounts on every render.
This implementation that I'm suggesting applies the optimization that you correctly applied to avoid additionals queries 👍 

I've also put up a codesandbox to show what happens when a component declaration is returned from a hook:

https://codesandbox.io/s/use-component-hook-pattern-b621kz?file=/src/App.tsx

You can see that whenever the parent re-render that component is unmounted (even though the parent is not unmounted) and this happen because the function defining the component is not stable.

Reply:
> Unfortunately when InstantSearch unmounts, there's nothing you can do to prevent all of its children from being unmounted

Totally, I'm aware of that. It turns out the extra re-rendering I was experiencing wasn't from the unmounting (`usePrevious` takes care of that), but from the `refresh` in the list of `useEffect` dependencies. See the pushed commit:
7c6ef136301a92a3b209f256cf66097e7daf3584

That said, I don't love that something *else* triggering it to re-render would result in double refreshes.